### ISSUE TITLE
Helm support: long term plan

### ISSUE THREAD

**Issue by natasha41575** [kubernetes-sigs/kustomize#4401]
There are many open helm-related issues in kustomize: https://github.com/kubernetes-sigs/kustomize/issues?q=is%3Aissue+is%3Aopen++label%3Aarea%2Fhelm+

Many of these issues are requesting new options in the helm chart generator configuration that will become fields to `helm template`. Other issues are related to problems with the `--enable-helm` flag and wanting more flexibility in how to enable helm. 

For some context, the `--enable-helm` flag exists and there are limited options in the plugin configuration because of a past security issue that has since been addressed. However, it is clear that the helm chart generator built into kustomize has limitations and we need a plan for long term support.

What we would like to do is introduce a new function in the new [krm-functions-registry](https://github.com/kubernetes-sigs/krm-functions-registry)  that can inflate a helm chart. This new function will be treated as a third party extension by kustomize so it will not need any special treatment. Additionally, because it will be a containerized function, there will be fewer security problems to worry about.

Once this function is ready, we can start recommending it to users and deprecate the existing helm-related fields in kustomize.

**Comment by james-callahan** [comment]
FWIW we don't consider krm functions usable in most of the environments we use kustomize.

On the other hand, I think the deprecation would help if the plugin is not going to get further support in kustomize itself. It might be worth directing people to https://github.com/mgoltzsche/khelm

**Comment by natasha41575** [comment]
Would you mind providing some more feedback for why krm functions are not usable for you? We would like to address friction points as we move toward more krm-function-oriented solutions and extensions.

**Comment by james-callahan** [comment]
> Would you mind providing some more feedback for why krm functions are not usable for you?

  - We require kustomize to be run on various end user devices that won't have docker installed (and/or aren't supported by docker)
  - We run kustomize from within CI pipelines that are unable to spawn a container
  - We run kustomize from within containers such as ArgoCD where they will not (and should not) have permissions to spawn a container

**Comment by khrisrichardson** [comment]
It's also worth mentioning that flux2's kustomize-controller doesn't natively support KRM config functions by way of plugin enablement, nor does it include the helm binary in the container image, and thus the current options are to [fork](https://github.com/fluxcd/kustomize-controller/issues/323#issuecomment-913767721) or pre-hydrate (which still sounds like it would be a challenge for @james-callahan until other runtimes are available for KRM config functions)

**Comment by snuggie12** [comment]
In #3536 I did get a special built version of `podman` working after renaming it to `docker` though would have preferred simply being able to name my runtime.

Our need matches what was previously stated since we use argocd; we did not want to install docker onto their repo-server. It would be a lot of work if the helm plugin was moved to a krm function and we'd likely stay on an older version or fork for quite some time.

Side note, in some prototyping we got around the whole `--enable-helm` by altering viper to accept it as an env var. Older versions of kustomize would ignore the env var and newer ones would process it correctly. I'm all in favor of getting rid of the flag if it is not needed but if that helps as a solution for any other flags I figured I'd mention it.

**Comment by KlavsKlavsen** [comment]
We are also looking to use this in argocd (as soon as kostumize gains OCI support - which was just released as stable with Helm 3.8.0) - so it would be bad, to limit it like this :(

**Comment by natasha41575** [comment]
We are looking at ways to support alternatives to docker, including containerized alternatives as well as simple and safe ways to run exec KRM functions. The latter would run very similarly to the existing builtin plugin.

**Comment by natasha41575** [comment]
Since I have been receiving several messages about this, I would like to reiterate again that we will be supporting the new KRM Helm function as executable, meaning that you would not need to run docker in order to use it. In terms of execution, this will be almost exactly the same as running the current builtin Helm plugin, except that the code lives elsewhere and is compiled separately.

**Comment by QuinnBast** [comment]
I noticed that the helm chart inflation has been moved to the `kubernetes-sigs` repository and now exists in both repositories. Which is the officially supported version, and how would I use `kubernetes-sigs` with kustomize?

**Comment by natasha41575** [comment]
> I noticed that the helm chart inflation has been moved to the kubernetes-sigs repository and now exists in both repositories. Which is the officially supported version, and how would I use kubernetes-sigs with kustomize?

Could you elaborate by what you mean by kubernetes-sigs repository? [kubernetes-sigs](https://github.com/kubernetes-sigs) is the org, and has many repositories.

**Comment by mwmahlberg** [comment]
I have noticed that the lack of the possibility to pass extra-args to helm makes it impossible to properly use helm charts that check capabilities, e.g. cert-manager's in-built Prometheus ServiceMonitor creation, see https://github.com/cert-manager/cert-manager/blob/9887baac3357beb1ad949f9b62f25dc3eea27fb7/deploy/charts/cert-manager/templates/servicemonitor.yaml#L1

My thoughts on that:

1. It makes sense for the upstream projects to check whether a resource is accepted by the target cluster, so the capability check is vital, especially when helm's [`--atomic`](https://helm.sh/docs/helm/helm_install/#options) flag is used (when helm is used directly, that is)
2. Speaking of which: One cannot pass the `--atomic` flag, either.
3. Imho, it does NOT make sense to implement the functionality and/or logic of helm's respective flags in kustomize -- that would couple the implementations tightly.
4. The easiest solution would have to (re-)introduce `extraArgs` to the `helmCharts` members and pass them verbatim to the function call

**Comment by james-callahan** [comment]
> I have noticed that the lack of the possibility to pass extra-args to helm makes it impossible to properly use helm charts that check capabilities

See https://github.com/kubernetes-sigs/kustomize/issues/3816

**Comment by mwmahlberg** [comment]
@james-callahan You sure are persistent in your efforts. +1

Some thoughts

- I guess this shows a wider problem: the divergence of the various means to deploy to k8s.
- One could argue that `kubectl` and by extension `kustomize` should be considered the canonical way and as such other tools need to adapt accordingly.
- I think, but I am not sure about it, that we have an edge case here: overriding the capabilities as determined by `Helm` is  only necessary if the logic in the template in question does come to a fail safe conclusion.

**Comment by KnVerey** [comment]
Please note that in light of slow progress on the KRM Function version, we've updated our mid-term plan around the current built-in. The new policy is outlined [in our docs](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/#long-term-support) and copied below for ease of reference.

Re: extra-args in particular: We used to have such a field, but we removed it in https://github.com/kubernetes-sigs/kustomize/pull/3784 because it represented a security risk. As such we will not be considering adding it again.

---

## Long term support

The helm chart inflation generator in kustomize is intended to be a limited subset of helm features to help with
getting started with kustomize, and we cannot support the entire helm feature set.

### The current builtin
For enhancements to the helm chart inflation generator feature, we will only support the following changes:

- bug fixes
- critical security issues
- additional fields that are analogous to flags passed to `helm template`, except for flags such as `post-renderer`
  that allow arbitrary commands to be executed

We will not add support for:

- private repository or registry authentication
- OCI registries
- other large features that increase the complexity of the feature and/or have significant security implications

### Future support
The next iteration of the helm inflation generator will take the form of a KRM function, which will have
no such restrictions on what types of features we can add and support. You can see more details in
the Helm support long term plan.

**Comment by jmmclean** [comment]
Its kind of a poor Developer Experience that the KRM repo is referenced for helm inflation, but no contributions have put on it for about 6 months - this is typically an indicator of a dying project in the open source world. Im specifically thinking of https://github.com/kubernetes-sigs/kustomize/issues/4335 where contributions were made by the community, but disregarded for the below option (clearly not an active project)

https://github.com/kubernetes-sigs/krm-functions-registry

**Comment by Jorricks** [comment]
Is this still an active plan? I am seeing no activity in the mentioned repository,  nor am I seeing any significant contributions in this repository to refactor the HELM setup.
IMO the HELM functionality works well, of course it has bugs, but all software does. Furthermore I think it's good to keep active contributions in this area until the entire refactor is complete.

Furthermore, I would like to extend upon the ideas of @mwmahlberg here. 
- I've come across [another implementation](https://github.com/jazzband/pip-tools#forwarding-options-to-pip) that will allow you to specify the args in the `kustomize` command, which would be the preferred way for me. The command would become something like; `kustomize build --enable-alpha-plugins --enable-helm --load-restrictor=LoadRestrictionsNone --helm-templated-args '--api-versions my-api-version' /path/to/deployment`

But I see the main reason for me to upgrade has been merged recently(I was looking for --api-versions support); https://github.com/kubernetes-sigs/kustomize/pull/4926

**Comment by james-callahan** [comment]
> As a downstream user, I'd like to be able to make use of Helm's `--atomic` flag

The `--atomic` flag doesn't make any sense in the context of `kustomize`.
Infact the usual way to apply kustomize is via `kustomize build | kubectl apply -f -` ==> the `kustomize` binary has no idea where it's output is going; whether it worked; or how to role it back.

**Comment by boldandbusted** [comment]
> > As a downstream user, I'd like to be able to make use of Helm's `--atomic` flag
> 
> The `--atomic` flag doesn't make any sense in the context of `kustomize`. Infact the usual way to apply kustomize is via `kustomize build | kubectl apply -f -` ==> the `kustomize` binary has no idea where it's output is going; whether it worked; or how to role it back.

Yep, I was an ignorant first-time Kustomize user when I posted that... and realize that using Kustomize as a way to apply Helm charts kind of removes a lot of the nice features of Helm as a K8s-equivalent of a OS package-manager.

**Comment by lindhe** [comment]
Is the plan to keep requiring the `--enable-helm` flag once this feature is complete, or can we get rid of it then?

**Comment by vladfr** [comment]
@KnVerey, I've reached this page from the docs. It doesn't seem that this Long-term plan for Helm is actually happening, as it's been a year since anyone has contributed to the KRM repo. It does seem like problems with the `--enable-helm` option are being addressed. 

Is it time to revisit the ideas outlined in this issue? Namely, there are PRs already done with support for private repos, OCI registry, and a few bugs. It would be great to reopen this. 

### Use-case: Cluster setup with ArgoCD

Helm in an Argo app can be a bit cumbersome to do with GitOps. There is a Chicken & Egg problem, where you need an App of Apps to initialize an Argo repo from Git, but Helm charts don't fit well with that. Argo UI also puts some limits in how it displays Helm charts too, with some nice advantages such as viewing values and overrides. But for the cluster admin use-case, it can get a bit cumbersome.

There is a case to be made that `kustomize` should be used instead. If we can nail support for Helm inflate within `kustomize` it would enable a nice use-case: **Apply a chart and kustomizations in the same Argo app**. This could be useful for cluster admins who want to consume org charts, or even public charts, and need to add or modify resources. 

Examples: 
 - install `cert-manager` and configure some ClusterIssuers
 - add Ingress resources for a Grafana chart

While this can be achieved in other ways, having `kustomize` apply these will enable a cleaner directory structure in the git repo, and enable the App-of-Apps pattern, which is a de-facto standard with ArgoCD and GitOps. (While I haven't fully tried it, the same can be said for Flux, although with Flux, the problem is smaller in scope because it doesn't extend to the UI)

**Comment by KlavsKlavsen** [comment]
For our open source project ( https://github.com/Obmondo/k8id ) - we only use helm charts and the 2.6+ feature with having values files in diff. repo - and in the very few instances where we need more, we have an app that simply points to static yaml files in the same "sepeate config repo" - or we extend helm chart and do upstream PR.. But helm + kustomize would give an extra option, which is definely not bad, but AFAIK kostumize's biggest strength is more the ability to handle patching of resources.. and a helm chart would "bark" at that afterwards? (unless argocd does helm template and runs kostumize locally - so it knows the end state - and there will then not be a diff.. I assume thats hot it would work). I still prefer upstream patch - but its definetely a good flexibility to have, to patch "resut of helm charts" - instead of the helm chart itself if necessary.

**Comment by GreenCappuccino** [comment]
> There is a case to be made that `kustomize` should be used instead. If we can nail support for Helm inflate within `kustomize` it would enable a nice use-case: **Apply a chart and kustomizations in the same Argo app**. This could be useful for cluster admins who want to consume org charts, or even public charts, and need to add or modify resources.

I actually do this, but instead of in the GitOps application, it's done at the pre-commit step. By computing the combined hash of all files in each directory to find which releases have changed, I can automatically generate the relevant Kustomize output, with inflated charts in a `build.yaml` file, onto which Kustomization patches/overlays can be applied onto. `build.yaml` is committed into version control.

It's not a very comprehensive solution, since it's in a single pre-commit python script, but I think that such a model has the potential to be compatible with both Argo and Flux, and also enable static-analysis based CI tooling, like for example to upgrade helm charts like Dependabot.

Here is how I manage my lab cluster (I'm planning to refactor the folder structure to support `kustomize localize`, but I haven't had the time): https://github.com/ShamrockTeam/kubernetes

**Comment by koba1t** [comment]
Hi all. I'm currently the subproject lead of kustomize.
I feel we need to improve helm support in this project because I think many users use this function in view of some issues and PRs.

In the first step, I plan to import helm lib in kustomize because it currently functions by directly running the `helm` binary on the user's computer. Other functions and improvements are added after that are completed.

Current status: I'm waiting for this PR to be merged.
https://github.com/helm/helm/pull/12725

**Comment by Skaronator** [comment]
Just a heads-up: the Helm PR has been closed, and it looks like we might not see a solution to that issue in the near future. It might be best to explore solutions within Kustomize instead of relying on Helm.

I also gave it a shot myself with a Helm feature implementation. Even though I’m not a Golang expert, I didn’t get any feedback from the dev team. I even put together a Helm improvement proposal [here](https://github.com/helm/community/pull/230) but didn’t get a response. So, I wouldn’t expect any updates from the Helm side anytime soon.

**Comment by GreenCappuccino** [comment]
Note that Kustomize supports KRM fn generators like the kpt helm chart rendering generator: https://catalog.kpt.dev/render-helm-chart/v0.1/
So you can use any arbitrary generator as long as it's containerized.

**Comment by Skaronator** [comment]
While KRM does exist, I think it's a flawed approach. Unlike [Helm plugins](https://helm.sh/docs/topics/plugins/) or [kubectl plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/), KRMs depend on a container runtime, which is usually unavailable in CI pipelines or CD tools such as ArgoCD.

I understand that supporting Helm involves a lot of overhead, and I can see why Kustomize might not want to handle Helm-related issues. Personally, I think it would be ideal if Kustomize adopted a plugin approach similar to kubectl or Helm and outsourced all Helm-related functionality.

**Comment by jashan-lco** [comment]
Standalone Kustomize (i.e. not backed-into kubectl) can do exec KRMs too: https://github.com/search?q=repo%3Akubernetes-sigs%2Fkustomize+%22config.kubernetes.io%2Ffunction%22+exec&type=code

You could grab the render-helm-chart KRM binary from the container image (or build it from source) and run kustomize with `--enable-exec`.

**Comment by QuinnBast** [comment]
It's pretty sad to see that this has been stalled for over 2 years, and that [we STILL can't login to private helm registries](https://github.com/kubernetes-sigs/kustomize/issues/4335) despite that being a very basic and necessary functionality to use helm at all...

Additionally, the docs for Kustomize's helm chart inflator [have been completely pulled](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/) and link to this github issue and the docs no longer explain the basic functionality that Kustomize currently supports through the helm inflation generator.

Edit: After browsing, I found the [actual docs for the helm inflation generator](https://kubectl.docs.kubernetes.io/references/kustomize/builtins/#_helmchartinflationgenerator_) but they are pretty buried, and the first result on google links to here which seems to indicate a lack of support for helm altogether...

**Comment by jeffmccune** [comment]
@QuinnBast 

FYI, [Holos](https://holos.run/docs/overview/) offers much of the same functionality as the helm chart inflator for the use case of kustomizing helm output to fully rendered manifests.  It runs fast locally and in CI with minimal dependencies, `helm` and `kubectl` or `kustomize` at present, and doesn't need a container runtime.

Helm is a first class citizen, as is Kustomize, and it's all glued together with [CUE](https://cuelang.org/) so you have another tool in the box if you'd rather not generate or transform data with helm, kustomize, or KRM functions.

Private registry authentication is supported and charts are automatically cached locally with support for vendoring them by simply committing them.

I'm the main author and maintainer and hang out in the CNCF slack among other places if you try it and get stuck I'd like to help.

**Comment by isarns** [comment]
Hi everyone,

I previously submitted a PR (https://github.com/kubernetes-sigs/kustomize/pull/5864) that removed the temporary Helm config and added support for pulling with the repository name. However, it was closed because Kustomize needs to ensure consistent behavior across all environments, avoiding different outcomes based on the setup.

From my understanding, Kustomize does not intend to support private Helm repositories or use existing Helm credentials stored in the Helm config directory—at least not without the temporary config directory.

Would it be possible to introduce a flag that explicitly allows Kustomize to use the system's Helm configuration or pass in credentials? This way, authentication could happen before running kustomize build.

I was particularly excited to see this https://github.com/helm/helm/pull/9760, as it could potentially resolve this issue if Kustomize leveraged the system’s Helm config.

**Comment by RobertDiebels** [comment]
> I get permission denied when trying to use `kustomize build --enable-helm` on public repos like oci://ghcr.io that allow anonymous authorization.
> 
> Is this the same issue? If not, apologies in advance, a search brought me here.

We have the same issue with a Harbor instance where a project has been set to public and private, both with the same authentication error.
We have this setup: ArgoCD -> Kustomize -> Helm -> Harbor

I've sifted through a lot of tickets, and it seems this has been an problem for a long time (well over 3 years now). All of which seem to hinge on the way Kustomize operates with Helm. Note that this is actually stated by Kustomize [ https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/#the-current-builtin ], but ArgoCD does not reference this issue, which is why I am here, instead of the argo repository.

Frankly I think this makes Kustomize unfit for usage with Helm at all. Private repositories are a main-stay in most enterprises, and tooling should support it. If it does not support private repositories then, attempting to do so should result in an understandable error, rather than having to read a authentication error thrown by Helm through Kustomize.

**Comment by titaneric** [comment]
For private helm chart, it's possible to pull it by providing [netrc](https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html#The-_002enetrc-file-1). You could take a look at [gitlab issue](https://gitlab.com/gitlab-org/gitlab/-/issues/350582) as well.
I have to admit that I have never try it before, but I adopt it in my company's argocd to reference remote git repo's kustomize something like the following.


```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
  - https://<my-git-repo>//configurations?ref=main
```

netrc:
```
machine <git-host> login <username> password <PAT>

```

**Comment by davinkevin** [comment]
🥁 Can't wait to see the coming decision as `kustomize` and `helm@v4` aren't very cooperative.

**Comment by lindhe** [comment]
@davinkevin Are you on the latest release of Kustomize? What's not working?

**Comment by mfredenhagen** [comment]
@davinkevin is my collegue and he encouraged me to chime in. Here is an issue I opened explaining the problem: https://github.com/kubernetes-sigs/kustomize/issues/6077

**Comment by davinkevin** [comment]
FYI, we had other issues, like this one: 

```
Error: accumulating resources: accumulation err='accumulating resources from 'chart': read /…/deployment/chart: is a directory': recursed accumulation of path '/…/deployment/chart': Error: unknown shorthand flag: 'c' in -c
: unable to run: 'helm version -c --short' with env=[HELM_CONFIG_HOME=/var/folders/2l/spr8j20n1j7453kjt35_3c0r0000gn/T/kustomize-helm-3027137771/helm HELM_CACHE_HOME=/var/folders/2l/spr8j20n1j7453kjt35_3c0r0000gn/T/kustomize-helm-3027137771/helm/.cache HELM_DATA_HOME=/var/folders/2l/spr8j20n1j7453kjt35_3c0r0000gn/T/kustomize-helm-3027137771/helm/.data] (is 'helm' installed?): exit status 1
error: no objects passed to apply
```

It works with `heml@v3` and `kustomize`, but it fails with `helm@v4`. @mfredenhagen could share more about it if needed 😇.

**Comment by koba1t** [comment]
@davinkevin 
`helm v4` is supported by kustomize v5.8.1 and above. Please use the latest version of kustomize.
https://github.com/kubernetes-sigs/kustomize/releases/tag/kustomize%2Fv5.8.1

**Comment by koba1t** [comment]
I think we need to refactor dependencies at helm before import helm pkg.
Importing it as-is would increase the size of the kustomize binary by about 40 MB, so this must be avoided.
https://github.com/helm/helm/issues/32468
